# Finding the Two Best Markets to Advertise in an E-learning Product

In this project, we'll aim to find the two best markets to advertise our product in — we're working for an e-learning company that offers courses on programming. Most of our courses are on web and mobile development, but we also cover many other domains, like data science, game development, etc.

# Understanding the Data

To avoid spending money on organizing a survey, we'll first try to make use of existing data to determine whether we can reach any reliable result.

One good candidate for our purpose is [freeCodeCamp's 2017 New Coder Survey](https://medium.freecodecamp.org/we-asked-20-000-people-who-they-are-and-how-theyre-learning-to-code-fff5d668969). [freeCodeCamp](https://www.freecodecamp.org/) is a free e-learning platform that offers courses on web development. Because they run [a popular Medium publication](https://medium.freecodecamp.org/) (over 400,000 followers), their survey attracted new coders with varying interests (not only web development), which is ideal for the purpose of our analysis.

The survey data is publicly available in [this GitHub repository](https://github.com/freeCodeCamp/2017-new-coder-survey). Below, we'll do a quick exploration of the `2017-fCC-New-Coders-Survey-Data.csv` file stored in the `clean-data` folder of the repository we just mentioned. We'll read in the file using the direct link [here](https://raw.githubusercontent.com/freeCodeCamp/2017-new-coder-survey/master/clean-data/2017-fCC-New-Coders-Survey-Data.csv).

In [1]:
# Reading the Data
import pandas as pd
direct_link = 'https://raw.githubusercontent.com/freeCodeCamp/2017-new-coder-survey/master/clean-data/2017-fCC-New-Coders-Survey-Data.csv'
fcc = pd.read_csv(direct_link,low_memory=0)

print(fcc.shape)
pd.options.display.max_columns = 150 # avoid truncated output
fcc.head()

(18175, 136)


Age  AttendedBootcamp  BootcampFinish  BootcampLoanYesNo BootcampName  \
0  27.0               0.0             NaN                NaN          NaN   
1  34.0               0.0             NaN                NaN          NaN   
2  21.0               0.0             NaN                NaN          NaN   
3  26.0               0.0             NaN                NaN          NaN   
4  20.0               0.0             NaN                NaN          NaN   

   BootcampRecommend  ChildrenNumber                 CityPopulation  \
0                NaN             NaN            more than 1 million   
1                NaN             NaN              less than 100,000   
2                NaN             NaN            more than 1 million   
3                NaN             NaN  between 100,000 and 1 million   
4                NaN             NaN  between 100,000 and 1 million   

   CodeEventConferences  CodeEventDjangoGirls  CodeEventFCC  CodeEventGameJam  \
0                   NaN                   NaN           NaN               NaN   
1                   NaN                   NaN           NaN               NaN   
2                   NaN                   NaN           NaN               NaN   
3                   NaN                   NaN           NaN               NaN   
4                   NaN                   NaN           NaN               NaN   

   CodeEventGirlDev  CodeEventHackathons  CodeEventMeetup  \
0               NaN                  NaN              NaN   
1               NaN                  NaN              NaN   
2               NaN                  1.0              NaN   
3               NaN                  NaN              NaN   
4               NaN                  NaN              NaN   

   CodeEventNodeSchool  CodeEventNone CodeEventOther  CodeEventRailsBridge  \
0                  NaN            NaN            NaN                   NaN   
1                  NaN            NaN            NaN                   NaN   
2                  1.0            NaN            NaN                   NaN   
3                  NaN            NaN            NaN                   NaN   
4                  NaN            NaN            NaN                   NaN   

   CodeEventRailsGirls  CodeEventStartUpWknd  CodeEventWkdBootcamps  \
0                  NaN                   NaN                    NaN   
1                  NaN                   NaN                    NaN   
2                  NaN                   NaN                    NaN   
3                  NaN                   NaN                    NaN   
4                  NaN                   NaN                    NaN   

   CodeEventWomenCode  CodeEventWorkshops       CommuteTime  \
0                 NaN                 NaN  15 to 29 minutes   
1                 NaN                 NaN               NaN   
2                 NaN                 NaN  15 to 29 minutes   
3                 NaN                 NaN  I work from home   
4                 NaN                 NaN               NaN   

             CountryCitizen               CountryLive  \
0                    Canada                    Canada   
1  United States of America  United States of America   
2  United States of America  United States of America   
3                    Brazil                    Brazil   
4                  Portugal                  Portugal   

               EmploymentField EmploymentFieldOther  \
0  software development and IT                  NaN   
1                          NaN                  NaN   
2  software development and IT                  NaN   
3  software development and IT                  NaN   
4                          NaN                  NaN   

                   EmploymentStatus EmploymentStatusOther  ExpectedEarning  \
0                Employed for wages                   NaN              NaN   
1  Not working but looking for work                   NaN          35000.0   
2                Employed for wages                   NaN          70000.0   

In [2]:
fcc.describe()

Age  AttendedBootcamp  BootcampFinish  BootcampLoanYesNo  \
count  15367.000000      17709.000000     1069.000000        1079.000000   
mean      27.691872          0.062002        0.699719           0.305839   
std        8.559239          0.241167        0.458594           0.460975   
min        0.000000          0.000000        0.000000           0.000000   
25%       22.000000          0.000000        0.000000           0.000000   
50%       26.000000          0.000000        1.000000           0.000000   
75%       32.000000          0.000000        1.000000           1.000000   
max       90.000000          1.000000        1.000000           1.000000   

       BootcampRecommend  ChildrenNumber  CodeEventConferences  \
count        1073.000000     2314.000000                1609.0   
mean            0.818267        1.832325                   1.0   
std             0.385805        0.972813                   0.0   
min             0.000000        1.000000                   1.0   
25%             1.000000        1.000000                   1.0   
50%             1.000000        2.000000                   1.0   
75%             1.000000        2.000000                   1.0   
max             1.000000        9.000000                   1.0   

       CodeEventDjangoGirls  CodeEventFCC  CodeEventGameJam  CodeEventGirlDev  \
count                 165.0        1708.0             290.0             297.0   
mean                    1.0           1.0               1.0               1.0   
std                     0.0           0.0               0.0               0.0   
min                     1.0           1.0               1.0               1.0   
25%                     1.0           1.0               1.0               1.0   
50%                     1.0           1.0               1.0               1.0   
75%                     1.0           1.0               1.0               1.0   
max                     1.0           1.0               1.0               1.0   

       CodeEventHackathons  CodeEventMeetup  CodeEventNodeSchool  \
count               1977.0           2472.0                446.0   
mean                   1.0              1.0                  1.0   
std                    0.0              0.0                  0.0   
min                    1.0              1.0                  1.0   
25%                    1.0              1.0                  1.0   
50%                    1.0              1.0                  1.0   
75%                    1.0              1.0                  1.0   
max                    1.0              1.0                  1.0   

       CodeEventNone  CodeEventRailsBridge  CodeEventRailsGirls  \
count         1076.0                 133.0                132.0   
mean             1.0                   1.0                  1.0   
std              0.0                   0.0                  0.0   
min              1.0                   1.0                  1.0   
25%              1.0                   1.0                  1.0   
50%              1.0                   1.0                  1.0   
75%              1.0                   1.0                  1.0   
max              1.0                   1.0                  1.0   

       CodeEventStartUpWknd  CodeEventWkdBootcamps  CodeEventWomenCode  \
count                 544.0                  547.0               481.0   
mean                    1.0                    1.0                 1.0   
std                     0.0                    0.0                 0.0   
min                     1.0                    1.0                 1.0   
25%                     1.0                    1.0                 1.0   
50%                     1.0                    1.0                 1.0   
75%                     1.0                    1.0                 1.0   
max                     1.0                    1.0                 1.0   

       CodeEventWorkshops  ExpectedEarning  FinanciallySupporting  \
count              1753.0      7200.000000           

# Checking for Sample Representativity

As we mentioned in the introduction, most of our courses are on web and mobile development, but we also cover many other domains, like data science, game development, etc. For the purpose of our analysis, we want to answer questions about a population of new coders that are interested in the subjects we teach. We'd like to know:

* Where are these new coders located.
* What locations have the greatest densities of new coders.
* How much money they're willing to spend on learning.

So we first need to clarify whether the data set has the right categories of people for our purpose. The `JobRoleInterest` column describes for every participant the role(s) they'd be interested in working in. If a participant is interested in working in a certain domain, it means that they're also interested in learning about that domain. So let's take a look at the frequency distribution table of this column and determine whether the data we have is relevant.

In [3]:
fcc.columns

Index(['Age', 'AttendedBootcamp', 'BootcampFinish', 'BootcampLoanYesNo',
       'BootcampName', 'BootcampRecommend', 'ChildrenNumber', 'CityPopulation',
       'CodeEventConferences', 'CodeEventDjangoGirls',
       ...
       'YouTubeFCC', 'YouTubeFunFunFunction', 'YouTubeGoogleDev',
       'YouTubeLearnCode', 'YouTubeLevelUpTuts', 'YouTubeMIT',
       'YouTubeMozillaHacks', 'YouTubeOther', 'YouTubeSimplilearn',
       'YouTubeTheNewBoston'],
      dtype='object', length=136)

In [5]:
# Frequency distribution table for 'JobRoleInterest'
fcc['JobRoleInterest'].value_counts(normalize = True) * 100

Full-Stack Web Developer                                                                                                                                                                                                                                                                    11.770595
  Front-End Web Developer                                                                                                                                                                                                                                                                    6.435927
  Data Scientist                                                                                                                                                                                                                                                                             2.173913
Back-End Web Developer                                                                                                

The information in the table above is quite granular, but from a quick scan it looks like:

* A lot of people are interested in web development (full-stack _web development_, front-end _web development_ and back-end _web development_).
* A few people are interested in mobile development.
* A few people are interested in domains other than web and mobile development.

It's also interesting to note that many respondents are interested in more than one subject. It'd be useful to get a better picture of how many people are interested in a single subject and how many have mixed interests. Consequently, in the next code block, we'll:

- Split each string in the `JobRoleInterest` column to find the number of options for each participant.
    - We'll first drop the null values because we can't split `Nan` values.
- Generate a frequency table for the variable describing the number of options.

In [6]:
# Split each string in the 'JobRoleInterest' column
interests_no_nulls = fcc['JobRoleInterest'].dropna()
splitted_interests = interests_no_nulls.str.split(',')

# Frequency table for the var describing the number of options
n_of_options = splitted_interests.apply(lambda x: len(x)) # x is a list of job options
n_of_options.value_counts(normalize = True).sort_index() * 100

1     31.650458
2     10.883867
3     15.889588
4     15.217391
5     12.042334
6      6.721968
7      3.861556
8      1.759153
9      0.986842
10     0.471968
11     0.185927
12     0.300343
13     0.028604
Name: JobRoleInterest, dtype: float64

It turns out that only 31.7% of the participants have a clear idea about what programming niche they'd like to work in, while the vast majority of students have mixed interests. But given that we offer courses on various subjects, the fact that new coders have mixed interest might be actually good for us.

The focus of our courses is on web and mobile development, so let's find out how many respondents chose at least one of these two options.